File: services/parcel-api/src/routes/parcels.js
'use strict';
const { Router } = require('express');
const pool = require('../db/client');

const router = Router();

const VALID_STATUSES = ['PENDING', 'IN_TRANSIT', 'OUT_FOR_DELIVERY', 'DELIVERED'];

function generateTrackingCode() {
  const chars = 'ABCDEFGHJKLMNPQRSTUVWXYZ23456789';
  let code = 'GL-';
  for (let i = 0; i < 6; i++) {
    code += chars[Math.floor(Math.random() * chars.length)];
  }
  return code;
}

// POST /parcels — créer un colis
router.post('/', async (req, res) => {
  try {
    const {
      senderName, recipientName, recipientEmail,
      recipientAddress, recipientLat, recipientLng, weightKg
    } = req.body;

    const trackingCode = generateTrackingCode();

    const result = await pool.query(
      `INSERT INTO parcels
         (tracking_code, sender_name, recipient_name, recipient_email,
          recipient_address, recipient_lat, recipient_lng, weight_kg)
       VALUES ($1,$2,$3,$4,$5,$6,$7,$8)
       RETURNING id, tracking_code, status, created_at`,
      [trackingCode, senderName, recipientName, recipientEmail,
       recipientAddress, recipientLat || null, recipientLng || null, weightKg || null]
    );

    const parcel = result.rows[0];

    await pool.query(
      `INSERT INTO delivery_events (parcel_id, event_type, description)
       VALUES ($1, 'CREATED', 'Colis créé')`,
      [parcel.id]
    );

    res.status(201).json(parcel);
  } catch (err) {
    console.error('[parcel-api] POST /parcels error:', err.message);
    res.status(500).json({ error: 'Internal server error' });
  }
});

// GET /parcels?trackingCode=GL-XXXX — recherche par code
router.get('/', async (req, res) => {
  const { trackingCode } = req.query;
  if (!trackingCode) {
    return res.status(400).json({ error: 'trackingCode query param is required' });
  }
  try {
    const result = await pool.query(
      'SELECT * FROM parcels WHERE tracking_code = $1',
      [trackingCode.toUpperCase()]
    );
    if (!result.rows.length) return res.status(404).json({ error: 'Parcel not found' });
    res.json(result.rows[0]);
  } catch (err) {
    res.status(500).json({ error: 'Internal server error' });
  }
});

// GET /parcels/:id — récupérer par UUID
router.get('/:id', async (req, res) => {
  try {
    const result = await pool.query(
      'SELECT * FROM parcels WHERE id = $1',
      [req.params.id]
    );
    if (!result.rows.length) return res.status(404).json({ error: 'Parcel not found' });
    res.json(result.rows[0]);
  } catch (err) {
    res.status(500).json({ error: 'Internal server error' });
  }
});

// PATCH /parcels/:id/status — mettre à jour le statut
router.patch('/:id/status', async (req, res) => {
  const { status, notifiedAt } = req.body;
  if (!VALID_STATUSES.includes(status)) {
    return res.status(400).json({ error: `Invalid status. Valid: ${VALID_STATUSES.join(', ')}` });
  }
  try {
    let query, values;
    if (notifiedAt) {
      query = `UPDATE parcels SET status=$1, notified_at=$2, updated_at=NOW()
               WHERE id=$3 RETURNING id, tracking_code, status, updated_at, notified_at`;
      values = [status, notifiedAt, req.params.id];
    } else {
      query = `UPDATE parcels SET status=$1, updated_at=NOW()
               WHERE id=$2 RETURNING id, tracking_code, status, updated_at`;
      values = [status, req.params.id];
    }

    const result = await pool.query(query, values);
    if (!result.rows.length) return res.status(404).json({ error: 'Parcel not found' });

    await pool.query(
      `INSERT INTO delivery_events (parcel_id, event_type, description)
       VALUES ($1, 'STATUS_CHANGE', $2)`,
      [req.params.id, `Statut changé en ${status}`]
    );

    res.json(result.rows[0]);
  } catch (err) {
    console.error('[parcel-api] PATCH status error:', err.message);
    res.status(500).json({ error: 'Internal server error' });
  }
});

module.exports = router;

File: services/parcel-api/src/routes/parcels.py
import random
import string

from fastapi import APIRouter, HTTPException, Query

from ..db.client import pool_manager, record_to_dict
from ..schemas import ParcelCreate, StatusUpdate

router = APIRouter()


def _gen_tracking_code() -> str:
    suffix = "".join(random.choices(string.ascii_uppercase + string.digits, k=6))
    return f"GL-{suffix}"


@router.post("", status_code=201)
async def create_parcel(body: ParcelCreate):
    pool = await pool_manager.get_pool()
    code = _gen_tracking_code()
    async with pool.acquire() as conn:
        row = await conn.fetchrow(
            """
            INSERT INTO parcels
              (tracking_code, sender_name, recipient_name,
               recipient_email, recipient_address, recipient_lat, recipient_lng)
            VALUES ($1,$2,$3,$4,$5,$6,$7)
            RETURNING *
            """,
            code,
            body.senderName,
            body.recipientName,
            body.recipientEmail,
            body.recipientAddress,
            body.recipientLat,
            body.recipientLng,
        )
        await conn.execute(
            "INSERT INTO delivery_events (parcel_id, event_type) VALUES ($1, 'CREATED')",
            row["id"],
        )
    return record_to_dict(row)


@router.get("")
async def list_parcels(trackingCode: str | None = Query(None)):
    pool = await pool_manager.get_pool()
    async with pool.acquire() as conn:
        if trackingCode:
            rows = await conn.fetch(
                "SELECT * FROM parcels WHERE tracking_code = $1", trackingCode.upper()
            )
            if not rows:
                raise HTTPException(status_code=404, detail="Parcel not found")
            return record_to_dict(rows[0])
        rows = await conn.fetch(
            "SELECT * FROM parcels ORDER BY created_at DESC LIMIT 50"
        )
    return [record_to_dict(r) for r in rows]


@router.get("/{parcel_id}")
async def get_parcel(parcel_id: str):
    pool = await pool_manager.get_pool()
    async with pool.acquire() as conn:
        row = await conn.fetchrow(
            "SELECT * FROM parcels WHERE id = $1::uuid", parcel_id
        )
    if not row:
        raise HTTPException(status_code=404, detail="Parcel not found")
    return record_to_dict(row)


@router.patch("/{parcel_id}/status")
async def update_status(parcel_id: str, body: StatusUpdate):
    pool = await pool_manager.get_pool()
    async with pool.acquire() as conn:
        async with conn.transaction():
            row = await conn.fetchrow(
                """
                UPDATE parcels
                SET status = $1, notified_at = COALESCE($2, notified_at), updated_at = NOW()
                WHERE id = $3::uuid
                RETURNING *
                """,
                body.status.value,
                body.notifiedAt,
                parcel_id,
            )
            if not row:
                raise HTTPException(status_code=404, detail="Parcel not found")
            await conn.execute(
                "INSERT INTO delivery_events (parcel_id, event_type) VALUES ($1, $2)",
                row["id"],
                body.status.value,
            )
    return record_to_dict(row)
